ts/hooks/use_stake.ts

```diff
@@ -65,110 +63,128 @@ export const useStake = (networkId: ChainId, providerState: ProviderState): UseS
-        setLoadingState(TransactionLoadingState.WaitingForSignature);
+    const executeWithData = useCallback(
+        async (data: string[]) => {
+            setLoadingState(TransactionLoadingState.WaitingForSignature);
 
-        const gasInfo = await backendClient.getGasInfoAsync();
+            const gasInfo = await backendClient.getGasInfoAsync();
 
-        const txPromise = stakingProxyContract
-            .batchExecute(data)
-            .awaitTransactionSuccessAsync({ from: ownerAddress, gasPrice: gasInfo.gasPriceInWei });
+            const txPromise = stakingProxyContract
+                .batchExecute(data)
+                .awaitTransactionSuccessAsync({ from: ownerAddress, gasPrice: gasInfo.gasPriceInWei });
 
-        await txPromise.txHashPromise;
-        setEstimatedTimeMs(gasInfo.estimatedTimeMs);
-        setLoadingState(TransactionLoadingState.WaitingForTransaction);
-        // tslint:disable:await-promise
-        const txResult = await txPromise;
-        setResult(txResult);
-        setLoadingState(TransactionLoadingState.Success);
-    }, [ownerAddress, stakingProxyContract]);
-
-    const depositAndStakeAsync = useCallback(async (stakePoolData: StakePoolData[]) => {
-        if (!stakePoolData || stakePoolData.length === 0 || isTxInProgress(loadingState)) {
-            return;
-        }
+            await txPromise.txHashPromise;
+            setEstimatedTimeMs(gasInfo.estimatedTimeMs);
+            setLoadingState(TransactionLoadingState.WaitingForTransaction);
+            // tslint:disable:await-promise
+            const txResult = await txPromise;
+            setResult(txResult);
+            setLoadingState(TransactionLoadingState.Success);
+        },
+        [ownerAddress, stakingProxyContract],
+    );
+
+    const depositAndStakeAsync = useCallback(
+        async (stakePoolData: StakePoolData[]) => {
+            if (!stakePoolData || stakePoolData.length === 0 || isTxInProgress(loadingState)) {
+                return;
+            }
+
+            const normalizedPoolData = normalizeStakePoolData(stakePoolData);
+
+            const totalStakeBaseUnits = normalizedPoolData.reduce(
+                (memo: BigNumber, { amountBaseUnits }: { amountBaseUnits: BigNumber }) => {
+                    return memo.plus(amountBaseUnits);
+                },
+                new BigNumber(0, 10),
+            );
+
+            const data: string[] = [
+                stakingContract.stake(totalStakeBaseUnits).getABIEncodedTransactionData(),
+                ..._.flatMap(normalizedPoolData, ({ poolId, amountBaseUnits }) => [
+                    // TODO(kimpers): only call finalizePool if poolStatsByEpoch returns feesCollected > 0
+                    // https://github.com/0xProject/0x-monorepo/blob/51ca3109eb29e1b03199e98587403fc61f8f2716/contracts/staking/contracts/src/sys/MixinFinalizer.sol#L93-L107
+                    stakingContract.finalizePool(poolId).getABIEncodedTransactionData(),
```

Comment:
does this cost gas ? if so, would this mean the first person to deposit or withdrawal in an epoch would have to pay a premium?

Reply:
Yes it would cost a bit more for the first person to deposit or withdraw after an epoch. If we or the pool operators don't finalize automatically.